stdlib/public/core/StringStorageBridge.swift

```diff
@@ -335,6 +373,464 @@ extension __SharedStringStorage {
+      self._switchRepresentationsIfInvalidUTF16(utf16Bytes)
+      
+      switch self {
+      case .native(var str):
+        str.append(String(decoding: utf16Bytes, as: UTF16.self))
+        self = .native(str)
+      case .utf16(var arr):
+        arr.append(contentsOf: utf16Bytes)
+        self = .utf16(arr)
+      }
+    }
+    
+    func mutableCopy() -> AnyObject {
+      switch self {
+      case .native(let str):
+        return _SwiftNSMutableString(str)
+      case .utf16(let arr):
+        return _SwiftNSMutableString(brokenContents: arr)
+      }
+    }
+    
+    func copy() -> AnyObject {
+      // While __StringStorage instances aren't immutable in general,
+      // mutations may only occur when instances are uniquely referenced.
+      // Therefore, it is safe to return self here; any outstanding Objective-C
+      // reference will make the instance non-unique.
+      switch self {
+      case .native(let str):
+        return str._bridgeToObjectiveCImpl()
+      case .utf16(let arr):
+        return _SwiftNSMutableString(brokenContents: arr)
+      }
+    }
+  }
+  
+  private var _contents: Contents
+  
+  internal init(_ str: String) {
+    _contents = .native(str)
+    super.init()
+    _internalInvariant(asString._guts.isFastUTF8)
+  }
+  
+  internal init(brokenContents arr: [UTF16.CodeUnit]) {
+    _contents = .utf16(arr)
+    super.init()
+  }
+  
+  final internal var isStringBacked: Bool {
+    @_effects(readonly) @inline(__always) get {
+      if case .native(_) = _contents {
+        return true
+      }
+      return false
+    }
+  }
+  
+  final internal var asString: String {
+    @_effects(readonly) @inline(__always) get {
+      return _contents.asString
+    }
+  }
+  
+  @inline(__always)
+  final func withFastUTF8<R>(
+    _ f: (UnsafeBufferPointer<UInt8>) throws -> R
+  ) rethrows -> R {
+    return try asString._guts.withFastUTF8(f)
+  }
+  
+  @inline(__always)
+  internal var start: UnsafePointer<UInt8>? {
+    return _contents.start
+  }
+  
+  @inline(__always)
+  internal var count: Int { return asString.utf8.count }
+  
+  @inline(__always)
+  final internal var isASCII: Bool {
+    return _contents.isASCII
+  }
+  
+  @objc(length)
+  final internal var UTF16Length: Int {
+    @_effects(readonly) get {
+      return _contents.UTF16Length
+    }
+  }
+
+  @objc(hash)
+  final internal var hash: UInt {
+    @_effects(readonly) get {
+      return _getCocoaHash()
+    }
+  }
+
+  @objc(characterAtIndex:)
+  @_effects(readonly)
+  final internal func character(at offset: Int) -> UInt16 {
+    return _contents.character(at: offset)
+  }
+
+  @objc(getCharacters:range:)
+  @_effects(releasenone)
+  final internal func getCharacters(
+    _ buffer: UnsafeMutablePointer<UInt16>, range aRange: _SwiftNSRange
+  ) {
+    switch _contents {
+    case .native(_):
+      _getCharacters(buffer, aRange)
+    case .utf16(let arr):
+      _precondition(aRange.location >= 0 && aRange.length >= 0,
```

Comment:
Not just `arr[range]._copyContents()` or something similar?

Reply:
oh, thank you, I thought I must be missing something here